unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1387 @@
+        /// complex handling of non-`'static` heaped data in general. Instead, it's manually
+        /// required to ensure this instance is accessed only via its associated Token for the
+        /// entire lifetime.
+        ///
+        /// This is intentionally left to be non-`const` to forbid unprotected sharing via static
+        /// variables among threads.
+        pub(super) fn new(value: V) -> Self {
+            Self(UnsafeCell::new(value))
+        }
+
+        /// Returns a mutable reference with its lifetime bound to the mutable reference of the
+        /// given token.
+        ///
+        /// In this way, any additional reborrow can never happen at the same time across all
+        /// instances of [`TokenCell<V>`] conceptually owned by the instance of [`Token<V>`] (a
+        /// particular thread), unless previous borrow is released. After the release, the used
+        /// singleton token should be free to be reused for reborrows.
+        pub(super) fn borrow_mut<'t>(&self, _token: &'t mut Token<V>) -> &'t mut V {
+            unsafe { &mut *self.0.get() }
+        }
+    }
+
+    // Safety: Access to TokenCell is assumed to be only from a single thread by proper use of
+    // Token once after TokenCell is sent to the thread from other threads; So, both implementing
+    // Send and Sync can be thought as safe.
+    //
+    // In other words, TokenCell is technicall still !Send and !Sync. But there should be no legal
+    // use happening which requires !Send or !Sync to avoid undefined behavior.
+    unsafe impl<V> Send for TokenCell<V> {}
+    unsafe impl<V> Sync for TokenCell<V> {}
+
+    /// A auxiliary zero-sized type to enforce aliasing rule to [`TokenCell`] via rust type system
+    ///
+    /// Token semantically owns a collection of `TokenCell` objects and governs the _unique_
+    /// existence of mutable access over them by requiring the token itself to be mutably borrowed
+    /// to get a mutable reference to the internal value of `TokenCell`.
+    // *mut is used to make this type !Send and !Sync
+    pub(super) struct Token<V: 'static>(PhantomData<*mut V>);
+
+    impl<V> Token<V> {
+        /// Returns the token to acquire a mutable reference to the inner value of [TokenCell].
+        ///
+        /// This is intentionally left to be non-`const` to forbid unprotected sharing via static
+        /// variables among threads.
+        ///
+        /// # Panics
+        ///
+        /// This function will `panic!()` if called multiple times with same type `V` from the same
+        /// thread to detect potential misuses.
+        ///
+        /// # Safety
+        ///
+        /// This method should be called exactly once for each thread at most to avoid undefined
+        /// behavior when used with [`Token`].
+        #[must_use]
+        pub(super) unsafe fn assume_exclusive_mutating_thread() -> Self {
+            thread_local! {
```

Comment:
> WRT the ShortCounter `Token`-gating; the reason we can't or do not want to use an `AtomicU32` is simply because we want to have checked operations to handle overflow explicitly?

`AtomicU32` is slow, because it always emits atomic operations, necessitating the given cpu core to go to the global bus ;) by the way, `std::cell::Cell<ShortCounter>` can be used without runtime cost here, thanks to `ShortCounter: Copy`. But i wanted to dog-food `TokenCell` and to use `TokenCell`'s more concise api than `Cell`, which is made possible due to its additional constraints than `Cell`.

Also, note that `blocked_page_count` will be removed for upcoming opt.

Reply:
> As a summary, ...
> 
> As far as I can tell, ...

also, this is correct. thanks for wrapping up your understanding.